File: rs2025/src/day01.rs
use std::fs;

pub fn day01() {
    let test_file = "src/inputs/day01.test.in";
    let main_file = "src/inputs/day01.in";

    part1(&test_file);
    part1(&main_file);

    part2(&test_file);
    part2(&main_file);
}

fn part1(filename: &str) -> u32 {
    let content = fs::read_to_string(filename).expect("should have been able to read");
    let rows = content.split("\n");

    let mut curr: i32 = 50;
    let mut count: u32 = 0;

    for row in rows {
        if row.starts_with("L") {
            let delta: &i32 = &row[1..].parse().unwrap();
            curr -= delta;
        } else if row.starts_with("R") {
            let delta: &i32 = &row[1..].parse().unwrap();
            curr += delta;
        }

        if curr % 100 == 0 {
            count += 1;
        }
    }

    println!("{count}");
    count
}

fn part2(filename: &str) -> i64 {
    let content = fs::read_to_string(filename).expect("should have been able to read");
    let rows = content.split("\n");

    let mut curr: i64 = 50;
    let mut count: i64 = 0;

    for row in rows {
        let prev = curr;

        if row.starts_with("L") {
            let delta: &i64 = &row[1..].parse().unwrap();
            curr -= delta;
            if curr % 100 == 0 {
                count += 1;
            }
            if prev % 100 == 0 {
                count -= 1;
            }
        } else if row.starts_with("R") {
            let delta: &i64 = &row[1..].parse().unwrap();
            curr += delta;
        }

        let zeroes: i64 = (floor(prev) - floor(curr)).abs();
        count += zeroes;
    }

    println!("{count}");
    count
}

fn floor(mut n: i64) -> i64 {
    let mut c = 0;
    if n < 0 {
        while n < 0 {
            n += 100;
            c -= 1;
        }
    } else {
        while n > 100 {
            n -= 100;
            c += 1;
        }
    }
    return c;
}

File: 2025/day01.py
# ruff: noqa: E731

from functools import reduce
from utils import file_to_list

INPUT_FILE_TEST = "inputs/day01.test.in"
INPUT_FILE = "inputs/day01.in"


def part1(filename: str) -> int:
    rows = file_to_list(filename)

    curr = 50
    count = 0

    for row in rows:
        delta = int(row[1:])
        if row[0] == "L":
            delta *= -1

        curr = (curr + delta) % 100

        if curr == 0:
            count += 1

    return count


part1ol = lambda filename: reduce(lambda x,y:(x[0]+y,x[1]+int((x[0]+y)%100 == 0)),(((-1)**"LR".index(x[0]))*int(x[1:]) for x in open(filename).read().split("\n")[:-1]),(50,0))[1]

print(part1(INPUT_FILE_TEST))
print(part1ol(INPUT_FILE_TEST))
print(part1(INPUT_FILE))
print(part1ol(INPUT_FILE))


def part2(filename: str) -> int:
    rows = file_to_list(filename)

    curr = 50
    count = 0

    for row in rows:
        delta = int(row[1:])
        if row[0] == "L":
            delta *= -1

        prev, curr = curr, curr + delta
        count += abs((prev // 100) - (curr // 100))
        if row[0] == "L" and curr % 100 == 0:
            count += 1
        if row[0] == "L" and prev % 100 == 0:
            count -= 1

    return count


part2ol = lambda filename: reduce(lambda x,y:(x[0]+y,(x[1]+abs((x[0]//100)-((x[0]+y)//100))+int(y<0 and (x[0]+y)%100 == 0)-int(y<0 and (x[0])%100 == 0))),(((-1)**"LR".index(x[0]))*int(x[1:]) for x in open(filename).read().split("\n")[:-1]),(50,0))[1]

print(part2(INPUT_FILE_TEST))
print(part2ol(INPUT_FILE_TEST))
print(part2(INPUT_FILE))
print(part2ol(INPUT_FILE))
